[WIP] Use BigTable Columns as range keys

gouthamve opened the pull request:
@tomwilkie Still have TODOs that I want to tackle in refactorings. Also I used SHA256 to hash, let me know if that's okay.

One curious thing to check is the prefix queries, I haven't found anything other than regex to fix that.

/cc @bboreham

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/weaveworks/cortex/753)
<!-- Reviewable:end -->

tomwilkie commented:
Thanks Goutham!  Will take a look tomorrow.

bboreham commented:
@Gouthamve please give the motivation for this change.

Gouthamve commented:
Sorry, explained everything in the commit message but forgot to post here too.

Motivation comes from: https://github.com/weaveworks/cortex/issues/714 This would mean accessing lesser number of rows per query.

tomwilkie commented on pkg/chunk/gcp/fixtures.go:
```diff
@@ -21,6 +21,8 @@ const (
 	srv  *bttest.Server
 	name string
+
+	version int
```
I'm generally less of a fan of "version" fields, and more of a fan of feature flags, mainly because feature fields are more self-explanatory.

tomwilkie commented on pkg/chunk/gcp/storage_client.go:
```diff
@@ -34,57 +35,82 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
-	// this format, so we need to do a proper migration.
-	rowKey := hashValue + separator + string(rangeValue)
+	hasher := sha256.New()
+	hasher.Write([]byte(hashValue))
```
SHA256 is pretty slow, and we don't need its cryptographic properties.  I'd have probably picked FNV (fast and pretty uniform), and then append the row key to final hashed value to guarantee uniqueness.

tomwilkie commented on pkg/chunk/gcp/storage_client.go:
```diff
@@ -109,61 +135,61 @@ func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch)
+
 	if len(query.RangeValuePrefix) > 0 {
-		rowRange = bigtable.PrefixRange(query.HashValue + separator + string(query.RangeValuePrefix))
+		rOpts = append(rOpts, bigtable.RowFilter(bigtable.ColumnFilter(string(query.RangeValuePrefix)+".*"))) // TODO: Check again and anchor.
```
Hmm yeah this is not particularly desirable.  @mbrukman do you know of a better way?

mbrukman replied:
@garye, what are your thoughts on this?

garye replied:
I'm a bit confused, this is doing a ColumnFilter with a value prefix?

gouthamve replied:
Yep. We want to filter by prefix.

garye replied:
I was just confused because this is setting up a filter on column names, but the RangeValuePrefix implies that the filter should be used for values instead.

tomwilkie replied:
This is just a Cortex-ism; our datamodel at this level is super simple, consisting of three-tuples of (hash value, range value, 'cell' value).  We always specify the hash value (mapped to the row key in bigtable), want to do range queries over the range value, and equality filtering on the 'cell' value.

Sorry for the confusion.

garye replied:
Makes sense, thanks for the explanation! 

I'm not sure there's a significantly better way to do this... It's probably worth doing this filtering on the bigtable side but, if the amount of data that would get filtered out by this ColumnFilter is small, consider measuring against doing it all client-side.

tomwilkie replied:
I wonder, would it be valid to apply the same trick @bcotton added to the row ranges; use range query from prefix to prefix+null ?

tomwilkie commented on pkg/chunk/gcp/storage_client.go:
```diff
@@ -109,61 +135,61 @@ func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch)
-	}, bigtable.RowFilter(bigtable.FamilyFilter(columnFamily)))
+	hasher := sha256.New()
+	hasher.Write([]byte(query.HashValue))
+	hashValue := string(hasher.Sum(nil))
```
Would be good to share this with the code to generate the row key for writes.

tomwilkie commented on pkg/chunk/gcp/storage_client.go:
```diff
@@ -109,61 +135,61 @@ func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch)
+
+	for i := range val {
+		val[i].Column = strings.TrimPrefix(val[i].Column, columnFamily+":")
+		// TODO: Hacky hacky ^
```
You could push this into the readbatch.RangeValue function.

tomwilkie commented on pkg/chunk/storage/storage_client_test.go:
```diff
@@ -88,6 +106,135 @@ func testStorageClientChunks(t *testing.T, client chunk.StorageClient) {
+			Name:             c.batch[0].tableName,
+			ProvisionedRead:  100000,
+			ProvisionedWrite: 100000,
+		}))
```
Not a huge fan of using the mock store to get the expected result; I think I'd prefer having the test case spell out explicitly what we expect to return.

Gouthamve replied:
Hmm, it actually increases the test verbosity by a fair share. I'd rather add test-cases to increase the confidence in the in-mem version, but it's your call.

tomwilkie commented:
Done a first pass; let me know when you've addressed the comments.

Also, we should test this out and get some performance results before we merge, its not guaranteed to be an improvement.

Gouthamve commented:
@tomwilkie PTAL. I've addressed most of the comments.

I've made it a single implementation as much as possible, but the `QueryPages` implementation differs enough to warrant a new type, I guess. I am not a big fan of `queryFn` as I think it hurts readability, for example, if it is:

```
func QueryPages() {
  return sc.queryFn()
}
```

Where is this queryFn passed? Do we want to have large fn definitions in the struct initialiser?

For the tests, I've done testing using the in-mem/Mock implementation for TSDB where the mock has tests to make sure it works as intended. I can do that, or just test the implementation independently. Your call.

bboreham commented:
What is the status of this PR?

tomwilkie commented:
We're testing it this week.

On Tue, Apr 3, 2018 at 10:44 AM Bryan Boreham <notifications@github.com>
wrote:

> What is the status of this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/weaveworks/cortex/pull/753#issuecomment-378192124>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbGhdTmZ6Vy2IuiPtRrK-sPIco3K9Nxks5tk0RpgaJpZM4StnrB>
> .
>

bboreham commented:
How did the testing go?

csmarchbanks commented:
Looking through this it appears the change is not backwards compatible right now? Did I miss something, or are there plans to use a -from flag?

Also, if you need some more testing done, FreshTracks would be happy to help out!

tomwilkie commented:
We dropped the ball on this:  We did some testing, and couldn't explain a difference (increase) in size in the bigtables.  Its still a WIP.

tomwilkie commented:
This is block on pushing the schema switching up a layer (to the chunk store) so we can have a migration path, as discussed on a recent community call.
